core/src/main/java/software/amazon/awssdk/auth/ContainerCredentialsProvider.java

```diff
@@ -0,0 +1,202 @@
+        }
+
+        @Override
+        public String disabledExceptionMessage() {
+            return String.format(
+                "Credentials cannot be loaded from ECS because the ECS credentials environment variable (%s) and system "
+                + "property (%s) are not set or cannot be accessed due to the security manager.",
+                AwsSystemSetting.AWS_CONTAINER_SERVICE_ENDPOINT.environmentVariable(),
+                AwsSystemSetting.AWS_CONTAINER_SERVICE_ENDPOINT.property());
+        }
+    }
+
+    public static class FullUriGenericCredentialsEndpointProvider implements CredentialsEndpointProvider {
+
+        private static Set<String> ALLOWED_HOSTS =
+            Collections.unmodifiableSet(new HashSet<>(Arrays.asList("localhost", "127.0.0.1")));
+
+        @Override
+        public URI endpoint() throws IOException {
+            URI uri = invokeSafely(() -> new URI(AwsSystemSetting.AWS_CONTAINER_CREDENTIALS_FULL_URI.getStringValueOrThrow()));
+            if (!ALLOWED_HOSTS.contains(uri.getHost())) {
+
+                throw new SdkClientException(
+                    String.format("The full URI (%s) contained withing environment variable %s has an invalid host. "
+                                  + "Host can only be one of [%s].",
+                                  uri,
+                                  AwsSystemSetting.AWS_CONTAINER_CREDENTIALS_FULL_URI.environmentVariable(),
+                                  ALLOWED_HOSTS.stream().collect(joining(","))));
+            }
+            return uri;
+        }
+
+        @Override
+        public Map<String, String> headers() {
+            return AwsSystemSetting.AWS_CONTAINER_AUTHORIZATION_TOKEN.getStringValue()
+                                                                     .filter(StringUtils::isNotBlank)
+                                                                     .map(t -> Collections.singletonMap("Authorization", t))
+                                                                     .orElseGet(Collections::emptyMap);
+        }
+
+        @Override
+        public boolean isEnabled() {
+            return AwsSystemSetting.AWS_CONTAINER_CREDENTIALS_FULL_URI.getStringValue().isPresent();
+        }
+
+        @Override
+        public String disabledExceptionMessage() {
+            return String.format(
+                "Credentials cannot be loaded from the container because the credentials environment variable (%s) and system "
+                + "property (%s) are not set or cannot be accessed due to the security manager.",
+                AwsSystemSetting.AWS_CONTAINER_CREDENTIALS_FULL_URI.environmentVariable(),
+                AwsSystemSetting.AWS_CONTAINER_CREDENTIALS_FULL_URI.property());
+        }
+    }
+
+    /**
+     *
+     */
+
+    /**
+     * A builder for creating a custom a {@link ContainerCredentialsProvider}.
+     */
+    public static final class Builder {
```

Comment:
Should this follow our normal builder pattern?

Reply:
You mean interface / implementation? Possibly though in reality it's never used outside of this package and I've made them package-private.